mne/utils/docs.py

```diff
@@ -874,10 +874,27 @@
 docdict['rank'] = """
 rank : None | dict | 'info' | 'full'
     This controls the rank computation that can be read from the
-    measurement info or estimated from the data. See ``Notes``
-    of :func:`mne.compute_rank` for details."""
-docdict['rank_None'] = docdict['rank'] + 'The default is None.'
-docdict['rank_info'] = docdict['rank'] + 'The default is "info".'
+    measurement info or estimated from the data.
+
+    :data:`python:None` (default)
+        The rank will be estimated from the data after proper scaling of
+        different channel types.
+    ``'info'``
+        The rank is inferred from ``info``. If data have been processed
+        with Maxwell filtering, the Maxwell filtering header is used.
+        Otherwise, the channel counts themselves are used.
+        In both cases, the number of projectors is subtracted from
+        the (effective) number of channels in the data.
+        For example, if Maxwell filtering reduces the rank to 68, with
+        two projectors the returned value will be 68.
+    ``'full'``
+        The rank is assumed to be full, i.e. equal to the
+        number of good channels. If a `~mne.Covariance` is passed, this can
+        make sense if it has been (possibly improperly) regularized without
+        taking into account the true data rank.
+"""
```

Comment:
missing explanation of `rank = dict(...)`

Reply:
Yes so what does this actually do? What is the use case? Looking at the code, it seems it simply skips the rank estimation  for the channel types that are present in the dict? Why do we need that?

Reply:
> What is the use case? Looking at the code, it seems it simply skips the rank estimation for the channel types that are present in the dict? Why do we need that?

Precisely the one I mentioned that I use [above](https://github.com/mne-tools/mne-python/pull/8707#discussion_r553328204): compute rank once on epochs to obtain a `rank` dict, then pass this `rank` dict to `compute_covariance`, `make_inverse_operator`, etc.

Reply:
Ok this **does** make sense for **those** functions, but clearly not for `compute_rank` – but we **do** have support there, part of the public API. I think what should be done is refactor `compute_rank` such that we move the implementation into a private function, which gets called by `compute_rank`, `make_inverse_operator` etc; and `compute_rank` wouldn't accept a dict anymore, while `make_inverse_operator` and friends would.

Reply:
This seems like a bit of overkill / overengineering to deal with a use case that isn't really causing people problems

Reply:
(and it's even possible someone knows their EEG rank but not MEG, but wants the result to have both because they have both, so doing `out = compute_rank(epochs, rank=dict(eeg=59))` might even be a valid use case for example)

Reply:
> (and it's even possible someone knows their EEG rank but not MEG, but wants the result to have both because they have both, so doing `out = compute_rank(epochs, rank=dict(eeg=59))` might even be a valid use case for example)

Honestly I think that **this** is over-engineered. Because my naive approach would have been:

```python
out = compute_rank(epochs.copy().pick_types(meg=True))
```

Anyway, I've added a doctoring now :)

Reply:
> Honestly I think that this is over-engineered.

Indeed, but the important difference is that this is (perhaps) over-engineered at the user's end, I'm trying to avoid unnecessary code churn and complications at our (MNE maintenance) end. Plus if it's a valid use case (even if over-engineered) then we don't want to break it just to simplify our docstrings a bit.